### PR TITLE
Enable logit masking for OpenAI Chat models.

### LINKED ISSUE #23: Logit bias for chat OpenAI models / support for cl100k_base tokenizer
Currently, lmql doesn't support logit bias-based constraints for `openai/gpt-3.5-turbo` and `openai/gpt-4`. However, the [OpenAI documentation](https://platform.openai.com/docs/api-reference/chat/create#chat/create-logit_bias) states that `logit_bias` is indeed supported by the API. The reason why it currently doesn't work might have something to do with the fact that these models use a different tokenizer, `cl100k_base`, which doesn't seem to be well documented. For example, for the "list of things not to forget when going to the sea" example, here's the current prompt generated by lmql:

```json
{
  "model": "gpt-4",
  "max_tokens": 32,
  "temperature": 0.8,
  "user": "lmql",
  "stream": true,
  "logit_bias": {
    "16012": 100,
    "16598": 100,
    "24541": 100
  },
  "messages": [
    {
      "role": "user",
      "content": "A list of things not to forget when going to the sea (not travelling): \n- Sunglasses \n-  Ur \n- "
    }
  ]
}
```
However, tokens `16012`, `16598` and `24541` in `cl100k_base` are `"Urls"`, `" Crusher"` and `".getType"`, which explains why the model generates a completion `"Ur"`. The solution might be to use the https://github.com/openai/tiktoken library:

```python
>>> import tiktoken
>>> enc = tiktoken.get_encoding("cl100k_base")
>>> enc.encode("Volleyball")
[53, 35619, 4047]
>>> enc.encode("Sunscreen")
[31192, 8337]
>>> enc.encode("Bathing Suite")
[33, 44661, 21652]
```
And indeed using logit biases
```json
{
    "53": 100,
    "31192": 100,
    "33": 100
}
```
I'm getting `{"content":"Sun"}`

### PR DESCRIPTION
**Pull request by kharvd**
The tokenizer seems to work, but something is definitely missing (it doesn't continue sampling after the first token, see screenshot)
image

Fixes #23

### CODE REVIEW

File: src/lmql/runtime/bopenai/openai_api.py [hunk 1]
```diff
@@ -58,16 +59,62 @@ async def complete(**kwargs):
...
+    def vocab_size(self):
+        return self.enc.max_token_value
+    
+    @property
+    def eos_token_id(self):
+        return self.enc.eot_token
+
+    @property
+    def bos_token_id(self):
+        return self.enc.encode_single_token("<|endofprompt|>")
```
Comment (kharvd) [inline src/lmql/runtime/bopenai/openai_api.py:99]: There is no `<BOS>` token in `cl100k_base` - from the brief look, it didn't seem like we're actually sending this to the API?

**Comment by lbeurerkellner** [comment]
Thanks, I will have a look. Interesting that they do not use BOS, but this should be easily removable.

**Comment by ogencoglu** [comment]
I can not get this working with `'openai/gpt-3.5-turbo'`

image

How did you manage getting an output @kharvd ?

**Comment by lbeurerkellner** [comment]
Closing this due to inactivity and the inability of the Chat API to perform partial completions. We may have some interesting alternative approaches to resolve this problem in the future, but I don't see an immediate solution.

**Pull request closed without merging**